packages/shared/src/contexts/UITaskManager.tsx

```diff
@@ -0,0 +1,99 @@
+import { defer, DeferTuple } from '@dimensiondev/kit'
+import { EMPTY_LIST } from '@masknet/shared-base'
+import { ComponentType, createContext, createElement, FC, PropsWithChildren, useMemo, useState } from 'react'
+import type { InjectedDialogProps } from './components'
+
+type ExtendPromise<T> = Promise<T> & {
+    close(): Promise<T>
+}
+interface ContextOptionsGeneric<T, R> {
+    show(options: T): ExtendPromise<R>
+}
+
+type BaseDialogProps = Pick<InjectedDialogProps, 'open' | 'onClose'>
+type CallableKeys<T> = keyof {
+    [K in keyof T as T[K] extends ((...args: any[]) => any) | undefined ? K : never]: any
+}
+
+/**
+ * Create a manager of small UI task sessions,
+ * which provide both a Context and a Provider.
+ */
+export const createUITaskManager = <
+    TaskOptions extends Record<string, any>,
+    Result,
+    Props extends BaseDialogProps,
+    ResolveProp extends CallableKeys<Props> & CallableKeys<TaskOptions>,
+>(
+    Component: ComponentType<Props>,
+    taskResolveField: ResolveProp,
+) => {
```

Comment:
I suggest the following changes:

- Remove `ExtendPromise`, and change to use `AbortController`.
- Remove `ResolveProp` type, and enforce the component to use a blessed name, e.g. `onSubmit(value: T): void`.

Therefore, I suggest changing the type signature to the below:

```ts
export function createUITaskManager<
    Opts extends { open: boolean, onClose(): void, onSubmit(value: T): void },
    T
>(component: React.ComponentType<Opts>): [
    useUI: () => (options: Options<Opts>, signal?: AbortSignal) => Promise<T>,
    provider: React.ComponentType<React.PropsWithChildren<{}>>
]
```

Usage:

```ts
const showDialog = useUI()

async function onClick() {
    const abort = new AbortController()
    const result = await showDialog({ title: 'abc' }, abort.signal)
}
```